File: packages/libs/SubstrateTypes/Sources/SubstrateTypes/MatrixF.swift
// MatrixF.swift
//
// Field-presence matrix F per cookbook § 6.1.
//
// F is a dense 2D array indexed by (field_index, bit_position),
// counting the number of rows in the estate where each
// (field, bit) is set. v0.36 has 36 fields × 6 bits = 216 cells.
//
// Storage: flat [Int64] of length 216, indexed
// `field_index * BITS_PER_FIELD + bit_position`.
//
// Update rule (cookbook § 6.1):
//   On row capture, for each (field, bit_position) where the
//   row's bitmap has the bit set, F[field, bit] += 1.
//   On row expunge of an active row, F[field, bit] -= 1.
//   On mutate, the delta is the difference between old and new
//   bit-presence states.
//
// Decay: F does NOT decay (cookbook § 6.8 table). It is a
// population statistic; halving it would lose information.

import Foundation

public struct MatrixF: Sendable, Equatable {

    // Phase 6 dedup: the 36×6 row layout has a canonical home in
    // RowBitmaps. MatrixF aliases these so the constants exist in
    // exactly one place per type system. (The values must match
    // RowBitmaps' by definition — they describe the same cookbook §2.3
    // layout — so an inequality here is a load-time crash, by design.)
    public static let fieldCount = RowBitmaps.fieldCount       // 36
    public static let bitsPerField = RowBitmaps.bitsPerField   // 6
    public static let cellCount = RowBitmaps.totalBits         // 216

    /// Flat storage. cells[field * 6 + bit] = count of rows where
    /// that bit is set in field's encoding.
    public private(set) var cells: [Int64]

    public init() {
        self.cells = [Int64](repeating: 0, count: Self.cellCount)
    }

    public init(cells: [Int64]) {
        precondition(cells.count == Self.cellCount,
                     "MatrixF requires exactly \(Self.cellCount) cells")
        self.cells = cells
    }

    // MARK: - Indexing

    @inlinable
    public static func cellIndex(field: Int, bit: Int) -> Int {
        precondition((0..<fieldCount).contains(field),
                     "field index \(field) out of range")
        precondition((0..<bitsPerField).contains(bit),
                     "bit position \(bit) out of range")
        return field * bitsPerField + bit
    }

    public subscript(field: Int, bit: Int) -> Int64 {
        get { return cells[Self.cellIndex(field: field, bit: bit)] }
        set { cells[Self.cellIndex(field: field, bit: bit)] = newValue }
    }

    // MARK: - Update rules

    /// Apply a row's bitmap-tier presence to the matrix.
    /// `delta` is +1 on capture, -1 on expunge. For mutate, call
    /// twice: once with -1 against the old bitmap, once with +1
    /// against the new bitmap.
    ///
    /// Phase 5 (decision 2026-05-28 §6.5): takes `BitVector216`
    /// (the dense 216-bit view) instead of a closure. Eliminates
    /// the per-call allocation of the closure and centralizes
    /// the field/bit layout in `RowBitmaps`.
    public mutating func applyRow(delta: Int64,
                                   bitVector: BitVector216) {
        guard delta != 0 else { return }
        for field in 0..<Self.fieldCount {
            for bit in 0..<Self.bitsPerField {
                if bitVector.bit(field: field, bit: bit) {
                    cells[Self.cellIndex(field: field, bit: bit)] &+= delta
                }
            }
        }
    }

    /// Total number of bit-presences recorded across all cells.
    /// Useful as a sanity check (should equal N_rows times the
    /// average bits-set-per-row).
    public var totalCount: Int64 {
        var sum: Int64 = 0
        for c in cells { sum &+= c }
        return sum
    }

    public mutating func reset() {
        for i in 0..<Self.cellCount { cells[i] = 0 }
    }

    // MARK: - Canonical wire form
    //
    // 216 × 8 bytes = 1728 bytes, all LE.

    public static let wireBytes = cellCount * 8

    public func writeWire(into bytes: inout [UInt8]) {
        for cell in cells {
            let u = UInt64(bitPattern: cell)
            for i in 0..<8 {
                bytes.append(UInt8((u >> (i * 8)) & 0xFF))
            }
        }
    }

    public static func readWire(_ bytes: [UInt8]) -> MatrixF? {
        guard bytes.count == wireBytes else { return nil }
        var cells = [Int64](repeating: 0, count: cellCount)
        for i in 0..<cellCount {
            var u: UInt64 = 0
            for j in 0..<8 {
                u |= UInt64(bytes[i * 8 + j]) << (j * 8)
            }
            cells[i] = Int64(bitPattern: u)
        }
        return MatrixF(cells: cells)
    }
}

// MARK: - Properties
//
//   linearity:   applyRow with delta=k followed by applyRow with
//                delta=-k restores the original matrix (modulo
//                overflow, which the &+= guards against).
//   monotone-on-capture: with only delta=+1 applications, every
//                cell is non-negative.
//   reset-idempotent: reset() applied twice equals reset() applied
//                once.
//
// MARK: - Cookbook references
//   § 6.1   Field-presence matrix F definition
//   § 6.8   Matrix decay table (F has half_life = None)
//   § A.1   Constitutional invariants I-1 through I-14

File: packages/libs/SubstrateTypes/rust/src/matrix_f.rs
// matrix_f.rs
//
// Field-presence matrix F per cookbook § 6.1. Mirror of
// Sources/SubstrateTypes/MatrixF.swift.

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MatrixF {
    /// Flat storage. cells[field * 6 + bit] = count of rows where
    /// that bit is set in field's encoding.
    cells: Vec<i64>,
}

impl MatrixF {
    pub const FIELD_COUNT: usize = 36;
    pub const BITS_PER_FIELD: usize = 6;
    pub const CELL_COUNT: usize = Self::FIELD_COUNT * Self::BITS_PER_FIELD; // 216
    pub const WIRE_BYTES: usize = Self::CELL_COUNT * 8;

    pub fn new() -> Self {
        Self {
            cells: vec![0; Self::CELL_COUNT],
        }
    }

    pub fn from_cells(cells: Vec<i64>) -> Self {
        assert_eq!(
            cells.len(),
            Self::CELL_COUNT,
            "MatrixF requires exactly {} cells",
            Self::CELL_COUNT
        );
        Self { cells }
    }

    pub fn cells(&self) -> &[i64] {
        &self.cells
    }

    // Indexing

    #[inline]
    pub fn cell_index(field: usize, bit: usize) -> usize {
        assert!(field < Self::FIELD_COUNT, "field {field} out of range");
        assert!(bit < Self::BITS_PER_FIELD, "bit {bit} out of range");
        field * Self::BITS_PER_FIELD + bit
    }

    pub fn get(&self, field: usize, bit: usize) -> i64 {
        self.cells[Self::cell_index(field, bit)]
    }

    pub fn set(&mut self, field: usize, bit: usize, value: i64) {
        self.cells[Self::cell_index(field, bit)] = value;
    }

    // Update rules

    /// Apply a row's bitmap-tier presence to the matrix.
    /// `delta` is +1 on capture, -1 on expunge. For mutate, call
    /// twice: once with -1 against the old bitmap, once with +1
    /// against the new bitmap.
    pub fn apply_row<F>(&mut self, delta: i64, bit_presence: F)
    where
        F: Fn(usize, usize) -> bool,
    {
        if delta == 0 {
            return;
        }
        for field in 0..Self::FIELD_COUNT {
            for bit in 0..Self::BITS_PER_FIELD {
                if bit_presence(field, bit) {
                    let idx = Self::cell_index(field, bit);
                    self.cells[idx] = self.cells[idx].wrapping_add(delta);
                }
            }
        }
    }

    pub fn total_count(&self) -> i64 {
        let mut sum: i64 = 0;
        for c in &self.cells {
            sum = sum.wrapping_add(*c);
        }
        sum
    }

    pub fn reset(&mut self) {
        for c in self.cells.iter_mut() {
            *c = 0;
        }
    }

    // Canonical wire form: 216 × 8 bytes, all LE.

    pub fn write_wire(&self, out: &mut Vec<u8>) {
        for c in &self.cells {
            out.extend_from_slice(&c.to_le_bytes());
        }
    }

    pub fn read_wire(bytes: &[u8]) -> Option<Self> {
        if bytes.len() != Self::WIRE_BYTES {
            return None;
        }
        let mut cells = vec![0i64; Self::CELL_COUNT];
        for i in 0..Self::CELL_COUNT {
            let mut buf = [0u8; 8];
            buf.copy_from_slice(&bytes[i * 8..i * 8 + 8]);
            cells[i] = i64::from_le_bytes(buf);
        }
        Some(Self { cells })
    }
}

impl Default for MatrixF {
    fn default() -> Self {
        Self::new()
    }
}

// Properties
//
//   linearity:   apply_row(delta=k, p) followed by apply_row(delta=-k, p)
//                restores the original matrix (modulo wrap; we use
//                wrapping_add which is overflow-safe).
//   monotone-on-capture: with only delta=+1, every cell is non-negative.
//   reset-idempotent: reset() applied twice equals reset() applied once.

#[cfg(test)]
mod tests {
    use super::*;

    fn always_set(_f: usize, _b: usize) -> bool {
        true
    }

    fn never_set(_f: usize, _b: usize) -> bool {
        false
    }

    #[test]
    fn fresh_matrix_zero() {
        let m = MatrixF::new();
        assert_eq!(m.total_count(), 0);
        assert_eq!(m.cells().len(), MatrixF::CELL_COUNT);
    }

    #[test]
    fn apply_row_all_set_capture() {
        let mut m = MatrixF::new();
        m.apply_row(1, always_set);
        assert_eq!(m.total_count(), MatrixF::CELL_COUNT as i64);
        for c in m.cells() {
            assert_eq!(*c, 1);
        }
    }

    #[test]
    fn apply_row_inverse() {
        let mut m = MatrixF::new();
        let row_pattern = |f: usize, b: usize| (f + b) % 2 == 0;
        m.apply_row(1, row_pattern);
        m.apply_row(-1, row_pattern);
        assert_eq!(m.total_count(), 0);
    }

    #[test]
    fn never_set_is_noop() {
        let mut m = MatrixF::new();
        m.apply_row(1, never_set);
        assert_eq!(m.total_count(), 0);
    }

    #[test]
    fn wire_round_trip() {
        let mut m = MatrixF::new();
        m.set(0, 0, 42);
        m.set(35, 5, -7);
        m.set(17, 3, 1_000_000);
        let mut bytes = Vec::new();
        m.write_wire(&mut bytes);
        assert_eq!(bytes.len(), MatrixF::WIRE_BYTES);
        let back = MatrixF::read_wire(&bytes).unwrap();
        assert_eq!(back, m);
    }
}
